target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c

```diff
@@ -1891,27 +1891,27 @@ int rtl838x_write_phy(u32 port, u32 page, u32 reg, u32 val)
 int rtl838x_read_mmd_phy(u32 port, u32 addr, u32 reg, u32 *val)
 {
 	u32 v;
+	u32 park_page;
 
 	mutex_lock(&smi_lock);
 
-	if (rtl838x_smi_wait_op(100000))
-		goto timeout;
-
-	sw_w32(1 << port, RTL838X_SMI_ACCESS_PHY_CTRL_0);
-	mdelay(10);
-
-	sw_w32_mask(0xffff0000, port << 16, RTL838X_SMI_ACCESS_PHY_CTRL_2);
+	reg = mdiobus_c45_regad(reg);
+	pr_debug("%s port %d, addr %d, reg %d\n", __func__, port, addr, reg);
+	sw_w32(port << 16, RTL838X_SMI_ACCESS_PHY_CTRL_2);
 
 	v = addr << 16 | reg;
 	sw_w32(v, RTL838X_SMI_ACCESS_PHY_CTRL_3);
 
-	/* mmd-access | read | cmd-start */
-	v = 1 << 1 | 0 << 2 | 1;
-	sw_w32(v, RTL838X_SMI_ACCESS_PHY_CTRL_1);
+	park_page = sw_r32(RTL838X_SMI_ACCESS_PHY_CTRL_1) & ((0x1f << 15) | 0x2);
+	/* mmd-access | read: BIT(2) not set | cmd-start */
+	v = BIT(1) | BIT(0);
```

Comment:
And TYPE_MMD is again set here. That's a bit redundant.

Reply:
That I can clean up :)